src/lsm/schema.zig

```diff
@@ -509,3 +521,137 @@ pub const TableData = struct {
         return schema.block_values_bytes_const(data_block)[0..used_bytes];
     }
 };
+
+/// A Manifest block's body is a SoA of Labels and TableInfos.
+/// Each Label/TableInfo pair is an "entry".
+// TODO Store timestamp (snapshot) in header.
+pub const Manifest = struct {
+    const entry_size = @sizeOf(TableInfo) + @sizeOf(Label);
+
+    pub const entry_count_max = @divFloor(block_body_size, entry_size);
+
+    const tables_size_max = entry_count_max * @sizeOf(TableInfo);
+    const labels_size_max = entry_count_max * @sizeOf(Label);
+
+    comptime {
+        assert(entry_count_max > 0);
+        assert(tables_size_max % @alignOf(Label) == 0);
+        assert(labels_size_max % @alignOf(Label) == 0);
+
+        // Bit 7 is reserved to indicate whether the event is an insert or remove.
+        assert(constants.lsm_levels <= std.math.maxInt(u7) + 1);
+
+        assert(@sizeOf(Label) == @sizeOf(u8));
+        assert(@alignOf(Label) == 1);
+
+        // All TableInfo's should already be 16-byte aligned because of the leading checksum.
+        const alignment = 16;
+        assert(alignment <= @sizeOf(vsr.Header));
+        assert(alignment == @alignOf(TableInfo));
+
+        // For keys { 8, 16, 24, 32 } all TableInfo's should be a multiple of the alignment.
+        // However, we still store Label ahead of TableInfo to save space on the network.
+        // This means we store fewer entries per manifest block, to gain less padding,
+        // since we must store entry_count_max of whichever array is first in the layout.
+        // For a better understanding of this decision, see schema.Manifest.size().
+        assert(@sizeOf(TableInfo) % alignment == 0);
+    }
+
+    /// Stored in every manifest block's header's `context` field.
+    pub const Context = extern struct {
+        entry_count: u32,
+        reserved: [12]u8 = .{0} ** 12,
+
+        comptime {
+            assert(@sizeOf(Context) == @sizeOf(u128));
+            assert(stdx.no_padding(Context));
+        }
+    };
+
+    /// See manifest.zig's TreeTableInfoType declaration for field documentation.
+    pub const TableInfo = extern struct {
+        /// All keys must fit within 32 bytes.
+        pub const KeyPadded = [32]u8;
+
+        checksum: u128,
+        address: u64,
+        tree_id: u16,
+        reserved: [6]u8 = .{0} ** 6,
+        snapshot_min: u64,
+        snapshot_max: u64,
+        key_min: KeyPadded,
+        key_max: KeyPadded,
+
+        comptime {
+            assert(@alignOf(TableInfo) == 16);
+            assert(stdx.no_padding(TableInfo));
+        }
+    };
+
+    pub const Label = packed struct(u8) {
+        level: u7,
+        event: enum(u1) { insert, remove },
+
+        comptime {
+            assert(@bitSizeOf(Label) == @sizeOf(Label) * 8);
+        }
+    };
+
+    entry_count: u32,
+
+    pub fn from(manifest_block: BlockPtrConst) Manifest {
+        const header = header_from_block(manifest_block);
```

Comment:
Nice! I don't think we ever use empty blocks. It might be redundant but what do you think of:
```zig
    assert(header.size >= @sizeOf(vsr.Header)); // Every block has a header.
    assert(header.size > @sizeOf(vsr.Header)); // Every block has a non-empty body.
```